atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/TransactionLocksManager.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.impl;
+
+import com.google.common.collect.ImmutableSet;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.v2.TimelockService;
+import java.io.Closeable;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import javax.annotation.concurrent.ThreadSafe;
+
+@ThreadSafe
+final class TransactionLocksManager implements Closeable {
```

Comment:
Let's add a java-doc here stating that this class is a utility that holds all the locks for the duration of a txn. As more locks are grabbed it keeps track of that.

Also, maybe SnapshotTransaction should not have access to `TimelockService#lock` without `TransactionLocksManager`.

Reply:
> Let's add a java-doc here stating that this class is a utility that holds all the locks for the duration of a txn. As more locks are grabbed it keeps track of that.

I'll add some docs.

> Also, maybe SnapshotTransaction should not have access to TimelockService#lock without TransactionLocksManager.
Interesting, lemme see what else do we use from TimelockService.

Reply:
I'll pass on that for now: there's a bunch of ceremony associated with locking, and you can lock in different ways (like we're currently adding locking a timestamp lease). I'd much rather not couple that to the manager, right now it's nicely scoped, and if we did this, we'd basically end up duplicating parts of APIs on TimelockService, for not much gain.

```        LockResponse lockResponse = timelockService.lock(
                request,
                ClientLockingOptions.builder()
                        .maximumLockTenure(
                                currentTransactionConfig.commitLockTenure().toJavaDuration())
                        .tenureExpirationCallback(() -> logCommitLockTenureExceeded(
                                lockDescriptors, currentTransactionConfig, stackTraceSnapshot))
                        .build());
```

Reply:
I am also bailing on the javadoc, I just renamed to TransactionLocksManager, should be fairly self explanatory?